src/distdl/backends/mpi/buffer.py

```diff
@@ -0,0 +1,185 @@
+        self.capacity = new_capacity
+        self.raw_buffer = new_buffer
+
+        # Loop over all existing views and recreate them in the new buffer.
+        new_views = dict()
+        for view_shape, view in self.views.items():
+            view_volume = np.prod(view_shape)
+            new_views[view_shape] = self.raw_buffer[:view_volume].reshape(view_shape)
+
+        self.views = new_views
+
+    def allocate_view(self, view_shape):
+        r"""Reserves a new view, with provided the shape, into the buffer.
+
+        A NumPy `ndarray` view of the 1D internal buffer with the desired
+        shape will be created.
+
+        Parameters
+        ----------
+        view_shape : iterable
+            Shape of the desired view.
+
+        """
+
+        # Ensure the shape is hashable
+        view_shape = tuple(view_shape)
+
+        self.get_view(view_shape)
+
+    def get_view(self, view_shape):
+        r"""Returns a view, with the provided shape, into the buffer.
+
+        A NumPy `ndarray` view of the 1D internal buffer with the desired
+        shape will be returned.  If the view does not exist it will be
+        created.
+
+        Parameters
+        ----------
+        view_shape : iterable
+            Shape of the desired view.
+
+        Returns
+        -------
+        The requested view.
+
+        """
+
+        # Ensure the shape is hashable
+        view_shape = tuple(view_shape)
+
+        # If there is already a view of this shape, then return it
+        if view_shape in self.views:
+            return self.views[view_shape]
+
+        # If new shape is larger than the buffer, expand the buffer
+        view_volume = np.prod(view_shape)
+        if view_volume > self.capacity:
+            self.expand(view_volume)
+
+        # Create the new view and return it
+        self.views[view_shape] = self.raw_buffer[:view_volume].reshape(view_shape)
+
+        return self.views[view_shape]
+
+
+class MPIBufferManager:
+    r"""NumPy (mpi4py compatible) implementation of an expandable buffer
+    manager.
+
+    Provides user interface for re-using expandable buffer objects.
+
+    Parameters
+    ----------
+    buffers : list
+        List of buffer objects
+
+    Attributes
+    ----------
+    buffers : list
+        List of buffer objects
+    """
+
+    def __init__(self):
+
+        self.buffers = list()
+
+    def request_buffers(self, n_buffers, dtype, **kwargs):
+        r"""Acquire a list of buffers of a specific dtype, creating them if
+        required.
+
+        Parameters
+        ----------
+        n_buffers : int
+            Number of buffers to acquire.
+        dtype : numpy.dtype
+            Data type of the requested buffers.
+
+        Returns
+        -------
+        List of `n_buffers` buffers with `dtype` data type.
+
+        """
+
+        # Extract a list of all existing buffers with matching dtype
+        dtype_buffers = [b for b in self.buffers if b.dtype == dtype]
```

Comment:
Super nitpicky, but this relationship between dtype and buffers would better be represented by a key-value map of some sort. Not sure if it's worth effort though, as the number of buffers should remain small enough that a search like this is O(1).

Reply:
I thought about this when implementing it and decided against it. The more I think on it the more I agree, though, so I'll make that change shortly.

Reply:
Done.